### ISSUE TITLE
metatype sequence scary error

### ISSUE THREAD

**Issue by Arnaz87** [nim-lang/Nim#4623]
I'm new to Nim and i'm making my first toy project. I'm not familiar with metatypes, but I needed a type placeholder for a type and using a `seq[void]` for a not implemented type compiled, so I kept it. But after a while I added some code.

``` nim
# Placeholder for a not yet implemented type
type Inst = void

# A few lines later
type
  Code = ref object of RootObj
    code: seq[Inst]
  State = ref object of RootObj
    code: Code

# The lines above were added at the beggining and compiled fine.
# The ones below were added later with a bunch of other code:
proc addState (code: Code) =
  var st = State() # <-- This line generates the error
  st.code = code
```

Compiling that generates this output:

```
Hint: used config file '/.../nim.cfg' [Conf]
Hint: system [Processing]
Hint: test [Processing]
test.nim(13, 6) Hint: 'test.addState(code: Code)' is declared but not used [XDeclaredButNotUsed]
CC: test
Error: execution of an external compiler program 'gcc -c  -w  -I/.../nim-0.14.2/lib -o /.../nimcache/test.o /.../nimcache/test.c' failed with exit code: 256

/.../nimcache/test.c: In function ‘TMP141’:
/.../nimcache/test.c:214:25: error: dereferencing pointer to incomplete type
  for (LOC1 = 0; LOC1 < a->Sup.len; LOC1++) {
                         ^
/.../nimcache/test.c: In function ‘testDatInit000’:
/.../test.c:389:24: error: ‘TGenericSeq’ undeclared (first use in this function)
 NTI95010.size = sizeof(TGenericSeq*);
                        ^
/.../nimcache/test.c:389:24: note: each undeclared identifier is reported only once for each function it appears in
/.../nimcache/test.c:389:36: error: expected expression before ‘)’ token
 NTI95010.size = sizeof(TGenericSeq*);
                                    ^
/.../nimcache/test.c:391:18: error: lvalue required as unary ‘&’ operand
 NTI95010.base = (&0);
                  ^
```

I'm new to Nim, now I understand (I believe) that that's not how metatypes should be used, but that is a weird, scary error, and it took me a while to find what caused it.

**Comment by Arnaz87** [comment]
This is a clearer example of the same error (I suppose):
```nim
type Broken = object
  sq: seq[void]

var o: Broken
```

And each of these I think are related but crashes differently, none of which is helpful:
```nim
var sq: seq[void]
var sq: seq[any]
var sq: seq[auto]
```

**Comment by andreaferretti** [comment]
I tried the second example, using `any`, `auto` or `void`, and in any case I get a nice error message `in.nim(2, 10) Error: 'void' is not a concrete type.` or similar.

Same for the first example: `in.nim(6, 14) Error: 'Inst' is not a concrete type.`

I guess this means the issue can be closed. @Arnaz87 can you confirm?

**Comment by Arnaz87** [comment]
Yes, Araq's commit fixed it.

With auto it still gives this weird error `Error: internal error: getTypeDescAux(tyAnything)` but it isn't propagating into the C compiler anymore, which was the reason of the big cryptic error.

**Comment by andreaferretti** [comment]
What version are you using? If you try it on https://play.nim-lang.org/ you will get a better error message (`Error: invalid type: 'Inst' in this context: 'proc (code: Code)'` for the first example and `Error: 'auto' is not a concrete type.` for the second one)

**Comment by Arnaz87** [comment]
Nim Compiler Version 0.17.0 (2017-05-22) [Linux: i386] on Arch Linux (Manjaro)

`var sq: seq[void]` gives `test.nim(1, 12) Error: 'void' is not a concrete type.`

`var sq: seq[auto]` gives `Error: internal error: getTypeDescAux(tyAnything)` (same with any, as it's an alias)

I don't know how to find the version of https://play.nim-lang.org/

**Comment by andreaferretti** [comment]
Are you sure it it not a problem with the Nim version in the Arch repository? On Ubuntu I get a nice error message both on devel and on v0.1.7

**Comment by Arnaz87** [comment]
That's weird, the release version is frozen so it can't be fixed by a side effect of a commit after it, which could be the case with devel but not with the Ubuntu package.

**Comment by Arnaz87** [comment]
The original issue is fixed long ago, but the other unfriendly error message still appears, even in https://play.nim-lang.org/, I just tested it (`var sq: seq[auto]`).

Should this be closed and a new issue opened?

**Comment by Araq** [comment]
> Should this be closed and a new issue opened?

I don't mind either way but please tell me the snippet that causes it.

**Comment by Arnaz87** [comment]
A file with this single line `var sq: seq[auto]` should reproduce it.

I would expect a nice error message like `Error: 'auto' is not a concrete type`, but instead gives a not so friendly `Error: internal error: getTypeDescAux(tyAnything)`

**Comment by Yardanico** [comment]
Now outputs:
```
code.nim(7, 14) Error: 'Inst' is not a concrete type.
```

**Comment by krux02** [comment]
The example has now a concrete error message.
